torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1210,10 +1210,17 @@ void ProcessGroupNCCL::heartbeatMonitor() {
   // we throw exception and make the whole process to be killed.
   const auto exitMsg = c10::str(
       logPrefix(),
-      "NCCL monitor thread timeout. Basically, this could ",
-      "be due to CUDA or NCCL calls being unexpectedly blocking, ",
-      "especially when your program enters a deadlock state in watchdog "
-      "or destructors. If you see this error, please file a bug to PyTorch.");
+      "ProcessGroupNCCL's watchdog got stuck for ",
+      heartbeatTimeoutInSec_,
+      "seconds without making progress in monitoring enqueued collectives. ",
+      "This typically indicates a NCCL/CUDA API hang blocking the watchdog, ",
+      "and could be triggered by another thread holding the GIL inside a ",
+      "CUDA api, or other deadlock-prone behaviors.",
+      "If you suspect the watchdog is not actually stuck and a longer timeout would help, ",
+      "you can either increase the timeout (TORCH_NCCL_HEARTBEAT_TIMEOUT_SEC) to a larger value "
+      "or disable the heartbeat monitor (TORCH_NCCL_ENABLE_MONITORING=0)."
+      "If either of aforementioned helps, feel free to file an issue to PyTorch about the short timeout "
+      "or false positive abort; otherwise, please report a watchdog hang bug to PyTorch.");
```

Comment:
> otherwise, please report a watchdog hang bug to PyTorch.

i think this part needs work, the rest looks great to me!

a watchdog hang is (usually?) not a pytorch bug.  So while it does make sense to contact us for help debugging, its not likely our bug.

"otherwise, please attempt to debug the hang by following X, Y, Z"

I think we should have a wiki for how to debug and triage such hangs.  And we can link to the wiki from this message.

Reply:
Yeah, that makes sense. But we need to help debug hang bug anyway that's why I put it in the log. (Let me add a TODO here. Maybe it's a good time for us to work on BE, documentation work now) We should include everything, from desync report, flight recorder, this monitor thread, etc.